storage/internal/benchmarks/client_pool.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/option"
+	htransport "google.golang.org/api/transport/http"
+	"google.golang.org/grpc"
+)
+
+type clientPool struct {
+	New     func() *storage.Client
+	clients []*storage.Client
+}
+
+func (p *clientPool) Get() *storage.Client {
+	if p.clients == nil {
+		p.clients = make([]*storage.Client, 0, opts.numWorkers)
+	}
+
+	if len(p.clients) > 0 {
+		c := p.clients[0]
+		p.clients = p.clients[1:]
+		return c
+	}
+
+	return p.New()
+}
+
+func (p *clientPool) Put(c *storage.Client) {
+	p.clients = append(p.clients, c)
+}
+
+// we can share clients as long as the app buffer sizes are constant
+var httpClients, gRPCClients *clientPool
+
+var nonBenchmarkingClients = clientPool{
+	New: func() *storage.Client {
+		// we don't care if it's grpc or http, so we don't need mutex
+		client, err := storage.NewClient(context.Background())
+		if err != nil {
+			log.Fatalf("storage.NewClient: %v", err)
+		}
+
+		return client
+	},
+}
+
+func initializeClientPools(opts *benchmarkOptions) func() {
+	httpClients = &clientPool{
+		New: func() *storage.Client {
+			client, err := initializeHTTPClient(context.Background(), opts.minWriteSize, opts.maxReadSize, opts.useDefaults)
+			if err != nil {
+				log.Fatalf("initializeHTTPClient: %v", err)
+			}
+
+			return client
+		},
+	}
+
+	gRPCClients = &clientPool{
+		New: func() *storage.Client {
+			client, err := initializeGRPCClient(context.Background(), opts.minWriteSize, opts.maxReadSize, opts.connPoolSize, opts.useDefaults)
+			if err != nil {
+				log.Fatalf("initializeGRPCClient: %v", err)
+			}
+			return client
+		},
+	}
+
+	return func() {
+		for _, c := range httpClients.clients {
+			c.Close()
+		}
+		for _, c := range gRPCClients.clients {
+			c.Close()
+		}
+	}
+}
+
+func canUseClientPool(opts *benchmarkOptions) bool {
+	return opts.useDefaults || (opts.maxReadSize == opts.minReadSize && opts.maxWriteSize == opts.minWriteSize)
```

Comment:
I don't really understand why this is the criterion for being able to use a client pool; could you add a comment?